In [11]:
import os
import pandas as pd
import datetime

import plotly.graph_objects as go
import plotly.express as px

In [12]:
df_load = pd.read_csv("/Users/nattapongkongkaew/Desktop/data_folder/load_data_15minresample_concat.csv")
df_pv = pd.read_csv("/Users/nattapongkongkaew/Desktop/data_folder/pv_data_15minresample_concat.csv")

In [13]:
# find mean of each column for each day
df_load['datetime'] = pd.to_datetime(df_load['datetime'])
df_load['date'] = df_load['datetime'].dt.date
df_load_mean = df_load.groupby(['date']).mean()
df_load_mean

/var/folders/pf/ns86k6hs2ljf3j58cd8dwmqh0000gn/T/ipykernel_5794/1344999366.py:4: FutureWarning:

The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.



Ia (A)     Ib (A)     Ic (A)   Pa (kW)   Pb (kW)   Pc (kW)  \
date                                                                        
2023-03-01  11.440447  11.632886  10.099835  1.473437  2.438124  2.425274   
2023-03-02   9.587602  17.038220  11.824135  1.879365  2.019245  3.529914   
2023-03-03   6.615404  13.661635   8.614466  1.319805  1.371056  2.840812   
2023-03-04   3.849860   5.613263   4.705525  0.967554  0.783085  1.188835   
2023-03-05   1.008731   3.509208   2.846061  0.919678  0.156675  0.738289   
...               ...        ...        ...       ...       ...       ...   
2024-02-24   2.984145   5.770482   9.328772  0.649458  1.245380  1.999765   
2024-02-25   3.062342   6.897564   4.840412  0.666713  1.527107  1.058180   
2024-02-27   7.078890  11.015588  18.460299  1.551649  2.411360  3.857881   
2024-02-28   7.502452  10.559761  22.905146  1.671114  2.277627  4.885130   
2024-02-29   8.349844  12.223056  20.214982  1.845823  2.624478  4.350500   

            Ptot (kW)  VAa (kVA)  VAb (kVA)  VAc (kVA)  ...  VARc (kVAR)  \
date                                                    ...                
2023-03-01   6.330926   0.000000   0.000000   0.000000  ...          NaN   
2023-03-02   7.403007   0.000000   0.000000   0.000000  ...          NaN   
2023-03-03   5.448367   0.000000   0.000000   0.000000  ...          NaN   
2023-03-04   2.910951   0.000000   0.000000   0.000000  ...          NaN   
2023-03-05   1.778006   0.000000   0.000000   0.000000  ...          NaN   
...               ...        ...        ...        ...  ...          ...   
2024-02-24   3.895237   0.679633   1.320803   2.112414  ...     0.667809   
2024-02-25   3.252643   0.696522   1.577777   1.097407  ...     0.291271   
2024-02-27   7.821057   1.613997   2.509510   4.177460  ...     1.578937   
2024-02-28   8.822866   1.710652   2.411391   5.177313  ...     1.689273   
2024-02-29   8.820120   1.902347   2.783255   4.580597  ...     1.482609   

            VARtot (kVAR)       PFa       PFb       PFc     PFtot       DFa  \
date                                                                          
2023-03-01            NaN       NaN       NaN       NaN       NaN       NaN   
2023-03-02            NaN       NaN       NaN       NaN       NaN       NaN   
2023-03-03            NaN       NaN       NaN       NaN       NaN       NaN   
2023-03-04            NaN       NaN       NaN       NaN       NaN       NaN   
2023-03-05            NaN       NaN       NaN       NaN       NaN       NaN   
...                   ...       ...       ...       ...       ...       ...   
2024-02-24       1.116536  0.915377  0.877587  0.924879  0.910522  0.957034   
2024-02-25       0.712407  0.916957  0.888888  0.965585  0.923291  0.958517   
2024-02-27       2.593597  0.924686  0.895767  0.904325  0.912737  0.952748   
2024-02-28       2.471866  0.961425  0.937994  0.973456  0.954056  0.976846   
2024-02-29       2.529358  0.971727  0.888105  0.945945  0.944013  0.989088   

                 DFb       DFc     DFtot  
date                                      
2023-03-01       NaN       NaN       NaN  
2023-03-02       NaN       NaN       NaN  
2023-03-03       NaN       NaN       NaN  
2023-03-04       NaN       NaN       NaN  
2023-03-05       NaN       NaN       NaN  
...              ...       ...       ...  
2024-02-24  0.912727  0.941696  0.936087  
2024-02-25  0.916852  0.977320  0.944485  
2024-02-27  0.914955  0.919204  0.928694  
2024-02-28  0.957805  0.975985  0.972138  
2024-02-29  0.896689  0.956865  0.946101  

[343 rows x 29 columns]

# Histogram of Load Consumption

In [14]:
# plot histogram where x is the mean of column Ptot and y is number of days using plotly_go

fig = go.Figure(data=[go.Histogram(x=df_load_mean['Ptot (kW)'],
                                   histnorm='probability'
                                   )])

fig.update_layout(
    title_text='Histogram of load column Ptot (kW) under 15 min resample',
    xaxis_title_text='Ptot (kW)',
    yaxis_title_text='Probability',
    bargap=0.2, # gap between bars of adjacent location coordinates.
    bargroupgap=0.1 # gap between bars of the same location coordinate.
)

fig.show()

# Histogram of Solar Generation

In [15]:
df_pv['datetime'] = pd.to_datetime(df_pv['datetime'])
df_pv['date'] = df_pv['datetime'].dt.date
df_pv['time'] = df_pv['datetime'].dt.time

In [16]:
df_pv

datetime  Ia (A)  Ib (A)  Ic (A)  Pa (kW)  Pb (kW)  Pc (kW)  \
0     2023-03-09 00:00:00    0.32    0.31    0.32  0.04119      0.0  0.05368   
1     2023-03-09 00:15:00    0.32    0.31    0.32  0.04119      0.0  0.05368   
2     2023-03-09 00:30:00    0.32    0.31    0.32  0.04119      0.0  0.05368   
3     2023-03-09 00:45:00    0.32    0.31    0.32  0.04119      0.0  0.05368   
4     2023-03-09 01:00:00    0.32    0.31    0.32  0.04119      0.0  0.05368   
...                   ...     ...     ...     ...      ...      ...      ...   
33787 2024-02-29 22:45:00    0.41    0.44    0.43  0.04339      0.0  0.11542   
33788 2024-02-29 23:00:00    0.41    0.44    0.43  0.04339      0.0  0.11542   
33789 2024-02-29 23:15:00    0.41    0.44    0.43  0.04339      0.0  0.11542   
33790 2024-02-29 23:30:00    0.41    0.44    0.43  0.04339      0.0  0.11542   
33791 2024-02-29 23:45:00    0.41    0.44    0.43  0.04339      0.0  0.11542   

       Ptot (kW)  VAa (kVA)  VAb (kVA)  VAc (kVA)  VAtot (kVA)        date  \
0        0.02756        0.0        0.0        0.0          0.0  2023-03-09   
1        0.02756        0.0        0.0        0.0          0.0  2023-03-09   
2        0.02756        0.0        0.0        0.0          0.0  2023-03-09   
3        0.02756        0.0        0.0        0.0          0.0  2023-03-09   
4        0.02756        0.0        0.0        0.0          0.0  2023-03-09   
...          ...        ...        ...        ...          ...         ...   
33787    0.06585        0.0        0.0        0.0          0.0  2024-02-29   
33788    0.06585        0.0        0.0        0.0          0.0  2024-02-29   
33789    0.06585        0.0        0.0        0.0          0.0  2024-02-29   
33790    0.06585        0.0        0.0        0.0          0.0  2024-02-29   
33791    0.06585        0.0        0.0        0.0          0.0  2024-02-29   

           time  
0      00:00:00  
1      00:15:00  
2      00:30:00  
3      00:45:00  
4      01:00:00  
...         ...  
33787  22:45:00  
33788  23:00:00  
33789  23:15:00  
33790  23:30:00  
33791  23:45:00  

[33792 rows x 14 columns]

In [17]:
start_time = pd.to_datetime('07:00:00').time()
end_time = pd.to_datetime('17:00:00').time()

# Filtering based on time within a range
df_pv_filtered = df_pv[(df_pv['datetime'].dt.time >= start_time) & (df_pv['datetime'].dt.time <= end_time)]
df_pv_filtered

datetime    Ia (A)    Ib (A)    Ic (A)   Pa (kW)  Pb (kW)  \
28    2023-03-09 07:00:00  0.313333  0.300939  0.263555  0.075130      0.0   
29    2023-03-09 07:15:00  0.405794  0.406492  0.410824  0.137692      0.0   
30    2023-03-09 07:30:00  0.669423  0.673530  0.680335  0.255284      0.0   
31    2023-03-09 07:45:00  0.940800  0.955916  0.959850  0.361650      0.0   
32    2023-03-09 08:00:00  1.018092  1.035521  1.038001  0.391870      0.0   
...                   ...       ...       ...       ...       ...      ...   
33760 2024-02-29 16:00:00  1.546373  1.555673  1.556901  0.576696      0.0   
33761 2024-02-29 16:15:00  1.181730  1.186337  1.188067  0.446157      0.0   
33762 2024-02-29 16:30:00  0.793717  0.789769  0.796156  0.298830      0.0   
33763 2024-02-29 16:45:00  0.491164  0.480605  0.487856  0.169324      0.0   
33764 2024-02-29 17:00:00  0.382256  0.356228  0.367783  0.110053      0.0   

        Pc (kW)  Ptot (kW)  VAa (kVA)  VAb (kVA)  VAc (kVA)  VAtot (kVA)  \
28     0.058025   0.070073        0.0        0.0        0.0          0.0   
29     0.046941   0.183452        0.0        0.0        0.0          0.0   
30     0.146757   0.402039        0.0        0.0        0.0          0.0   
31     0.242320   0.603829        0.0        0.0        0.0          0.0   
32     0.270983   0.662723        0.0        0.0        0.0          0.0   
...         ...        ...        ...        ...        ...          ...   
33760  0.444321   1.020848        0.0        0.0        0.0          0.0   
33761  0.321836   0.768417        0.0        0.0        0.0          0.0   
33762  0.188532   0.487436        0.0        0.0        0.0          0.0   
33763  0.076933   0.246148        0.0        0.0        0.0          0.0   
33764  0.030765   0.136451        0.0        0.0        0.0          0.0   

             date      time  
28     2023-03-09  07:00:00  
29     2023-03-09  07:15:00  
30     2023-03-09  07:30:00  
31     2023-03-09  07:45:00  
32     2023-03-09  08:00:00  
...           ...       ...  
33760  2024-02-29  16:00:00  
33761  2024-02-29  16:15:00  
33762  2024-02-29  16:30:00  
33763  2024-02-29  16:45:00  
33764  2024-02-29  17:00:00  

[14432 rows x 14 columns]

In [18]:
colors = px.colors.qualitative.Bold
fig = go.Figure()

for i, date in enumerate(df_pv_filtered['datetime'].dt.date.unique()):
    dff = df_pv_filtered[df_pv_filtered['datetime'].dt.date == date]
    fig.add_trace(go.Scatter(x=dff['time'],
                             y=dff['Ptot (kW)'],
                             mode='lines',
                             name=str(date),
                             line=dict(color=colors[int(i%len(colors))], width=1)
                             ))                            
    
fig.update_layout(title="<b>PV Power Generation<b>",
                  title_y = 0.95,
                  legend=dict(orientation="h",
                              yanchor="bottom",
                              y=1,
                              xanchor="right",
                              x=0.9
                             ))
# Show less frequent x-axis ticks
rs_length = 15 # in minute
fig.update_xaxes(dtick=int(60/rs_length))
fig.update_yaxes(title="Power (kW)")


In [19]:
# find mean of each column for each day
df_pv_mean = df_pv_filtered.groupby(['date']).mean()
df_pv_mean

/var/folders/pf/ns86k6hs2ljf3j58cd8dwmqh0000gn/T/ipykernel_5794/2009348549.py:2: FutureWarning:

The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.



Ia (A)    Ib (A)    Ic (A)   Pa (kW)  Pb (kW)   Pc (kW)  \
date                                                                    
2023-03-09  2.914897  2.955871  2.952926  1.066853      0.0  0.904490   
2023-03-10  2.742702  2.781129  2.778713  1.007833      0.0  0.849131   
2023-03-12  2.664832  2.700301  2.694720  0.973242      0.0  0.822429   
2023-03-13  2.829776  2.866607  2.864938  1.030981      0.0  0.873046   
2023-03-15  3.665682  3.718373  3.712831  1.330343      0.0  1.151484   
...              ...       ...       ...       ...      ...       ...   
2024-02-25  1.857524  1.866061  1.864621  0.677239      0.0  0.545666   
2024-02-26  1.441667  1.447965  1.448214  0.530186      0.0  0.407549   
2024-02-27  1.717574  1.728269  1.727472  0.629889      0.0  0.497879   
2024-02-28  1.609900  1.614841  1.612388  0.585012      0.0  0.461425   
2024-02-29  1.665173  1.667609  1.665239  0.604014      0.0  0.478802   

            Ptot (kW)  VAa (kVA)  VAb (kVA)  VAc (kVA)  VAtot (kVA)  
date                                                                 
2023-03-09   1.969770        0.0        0.0        0.0          0.0  
2023-03-10   1.854903        0.0        0.0        0.0          0.0  
2023-03-12   1.794260        0.0        0.0        0.0          0.0  
2023-03-13   1.901129        0.0        0.0        0.0          0.0  
2023-03-15   2.481197        0.0        0.0        0.0          0.0  
...               ...        ...        ...        ...          ...  
2024-02-25   1.219743        0.0        0.0        0.0          0.0  
2024-02-26   0.934482        0.0        0.0        0.0          0.0  
2024-02-27   1.124713        0.0        0.0        0.0          0.0  
2024-02-28   1.043033        0.0        0.0        0.0          0.0  
2024-02-29   1.079859        0.0        0.0        0.0          0.0  

[352 rows x 11 columns]

In [20]:
# plot histogram where x is the mean of column Ptot and y is number of days using plotly_go

fig = go.Figure(data=[go.Histogram(x=df_pv_mean['Ptot (kW)']
                                   )])

fig.update_layout(
    title_text='Histogram of pv column Ptot (kW) under 15 min resample',
    xaxis_title_text='Ptot (kW)',
    yaxis_title_text='Probability',
    bargap=0.2, # gap between bars of adjacent location coordinates.
    bargroupgap=0.1 # gap between bars of the same location coordinate.
)

fig.show()

# Categorize Load and Solar

## Load

### Pload < 4 kW => low_load
### 4 kW <= Pload < 6 kW => medium_load
### 6 kW <= Pload => high_load

## Solar

### Ppv <= 1.5 kW => low_solar
### 1.5 kW < Ppv => high_solar

In [21]:
def categorize_load(Pload):
    left_pivot = 4
    right_pivot = 6
    if Pload < left_pivot:
        load_type = 'low_load'
    elif left_pivot <= Pload and Pload < right_pivot:
        load_type = 'medium_load'
    elif right_pivot <= Pload:
        load_type = 'high_load'
        
    return load_type

In [22]:
def categorize_pv(Ppv):
    pivot = 1.5
    if Ppv <= pivot:
        solar_type = 'low_solar'
    elif pivot < Ppv:
        solar_type= 'high_solar'
        
    return solar_type

In [23]:
df_load_mean['load_type'] = df_load_mean['Ptot (kW)'].apply(categorize_load)
df_pv_mean['solar_type'] = df_pv_mean['Ptot (kW)'].apply(categorize_pv)

In [24]:
df_load_mean = df_load_mean.reset_index()
df_load_type = df_load_mean[['date', 'load_type']]
df_load_type

date    load_type
0    2023-03-01    high_load
1    2023-03-02    high_load
2    2023-03-03  medium_load
3    2023-03-04     low_load
4    2023-03-05     low_load
..          ...          ...
338  2024-02-24     low_load
339  2024-02-25     low_load
340  2024-02-27    high_load
341  2024-02-28    high_load
342  2024-02-29    high_load

[343 rows x 2 columns]

In [25]:
df_pv_mean

Ia (A)    Ib (A)    Ic (A)   Pa (kW)  Pb (kW)   Pc (kW)  \
date                                                                    
2023-03-09  2.914897  2.955871  2.952926  1.066853      0.0  0.904490   
2023-03-10  2.742702  2.781129  2.778713  1.007833      0.0  0.849131   
2023-03-12  2.664832  2.700301  2.694720  0.973242      0.0  0.822429   
2023-03-13  2.829776  2.866607  2.864938  1.030981      0.0  0.873046   
2023-03-15  3.665682  3.718373  3.712831  1.330343      0.0  1.151484   
...              ...       ...       ...       ...      ...       ...   
2024-02-25  1.857524  1.866061  1.864621  0.677239      0.0  0.545666   
2024-02-26  1.441667  1.447965  1.448214  0.530186      0.0  0.407549   
2024-02-27  1.717574  1.728269  1.727472  0.629889      0.0  0.497879   
2024-02-28  1.609900  1.614841  1.612388  0.585012      0.0  0.461425   
2024-02-29  1.665173  1.667609  1.665239  0.604014      0.0  0.478802   

            Ptot (kW)  VAa (kVA)  VAb (kVA)  VAc (kVA)  VAtot (kVA)  \
date                                                                  
2023-03-09   1.969770        0.0        0.0        0.0          0.0   
2023-03-10   1.854903        0.0        0.0        0.0          0.0   
2023-03-12   1.794260        0.0        0.0        0.0          0.0   
2023-03-13   1.901129        0.0        0.0        0.0          0.0   
2023-03-15   2.481197        0.0        0.0        0.0          0.0   
...               ...        ...        ...        ...          ...   
2024-02-25   1.219743        0.0        0.0        0.0          0.0   
2024-02-26   0.934482        0.0        0.0        0.0          0.0   
2024-02-27   1.124713        0.0        0.0        0.0          0.0   
2024-02-28   1.043033        0.0        0.0        0.0          0.0   
2024-02-29   1.079859        0.0        0.0        0.0          0.0   

            solar_type  
date                    
2023-03-09  high_solar  
2023-03-10  high_solar  
2023-03-12  high_solar  
2023-03-13  high_solar  
2023-03-15  high_solar  
...                ...  
2024-02-25   low_solar  
2024-02-26   low_solar  
2024-02-27   low_solar  
2024-02-28   low_solar  
2024-02-29   low_solar  

[352 rows x 12 columns]

In [26]:
df_pv_mean = df_pv_mean.reset_index()
df_pv_type = df_pv_mean[['date', 'solar_type']]
df_pv_type

date  solar_type
0    2023-03-09  high_solar
1    2023-03-10  high_solar
2    2023-03-12  high_solar
3    2023-03-13  high_solar
4    2023-03-15  high_solar
..          ...         ...
347  2024-02-25   low_solar
348  2024-02-26   low_solar
349  2024-02-27   low_solar
350  2024-02-28   low_solar
351  2024-02-29   low_solar

[352 rows x 2 columns]

# Batch

In [27]:
load = pd.read_csv('data/load_data_15minresample_concat.csv',parse_dates=[0]).rename(columns={'Pa (kW)':'PLa',
                                                                                         'Pb (kW)':'PLb',
                                                                                         'Pc (kW)':'PLc',
                                                                                         'Ptot (kW)':'PLtot'	
                                                                                         })
pv = pd.read_csv('data/pv_data_15minresample_concat.csv',parse_dates=[0]).rename(columns={'Pa (kW)':'PVa',	
                                                                                     'Pb (kW)':'PVb',
                                                                                     'Pc (kW)':'PVc',
                                                                                     'Ptot (kW)':'PVtot'	
                                                                                     })

In [28]:
pv.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33792 entries, 0 to 33791
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   datetime     33792 non-null  datetime64[ns]
 1   Ia (A)       33792 non-null  float64       
 2   Ib (A)       33792 non-null  float64       
 3   Ic (A)       33792 non-null  float64       
 4   PVa          33792 non-null  float64       
 5   PVb          33792 non-null  float64       
 6   PVc          33792 non-null  float64       
 7   PVtot        33792 non-null  float64       
 8   VAa (kVA)    33792 non-null  float64       
 9   VAb (kVA)    33792 non-null  float64       
 10  VAc (kVA)    33792 non-null  float64       
 11  VAtot (kVA)  33792 non-null  float64       
dtypes: datetime64[ns](1), float64(11)
memory usage: 3.1 MB


In [29]:
load.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32899 entries, 0 to 32898
Data columns (total 30 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   datetime       32899 non-null  datetime64[ns]
 1   Ia (A)         32899 non-null  float64       
 2   Ib (A)         32899 non-null  float64       
 3   Ic (A)         32899 non-null  float64       
 4   PLa            32899 non-null  float64       
 5   PLb            32899 non-null  float64       
 6   PLc            32899 non-null  float64       
 7   PLtot          32899 non-null  float64       
 8   VAa (kVA)      32899 non-null  float64       
 9   VAb (kVA)      32899 non-null  float64       
 10  VAc (kVA)      32899 non-null  float64       
 11  VAtot (kVA)    32899 non-null  float64       
 12  In (A)         26976 non-null  float64       
 13  Iavg (A)       26976 non-null  float64       
 14  Va (V)         26976 non-null  float64       
 15  Vb (V)         2697

In [30]:
start = max(load['datetime'].min(),pv['datetime'].min()) 
end = start + pd.Timedelta(9, unit='D')

In [31]:
#for 15 min sampling

num_day = 4 # unit : day
resolution = 15 # unit : min
num_interval_per_day = int(24*60/resolution)
num_total_interval = num_day * num_interval_per_day

# there are six types of batch dataset
type1 = 0
type2 = 0
type3 = 0
type4 = 0
type5 = 0
type6 = 0

start = max(load['datetime'].min(),pv['datetime'].min()) 
end = min(load['datetime'].max(),pv['datetime'].max())

dataset_detail = []
for i in range((end - start).days):
    today = start + pd.Timedelta(i,unit='D')
    next_4_days = today + pd.Timedelta(4,unit='D')
    load_batch = load[(today <=load['datetime']) & (load['datetime'] < next_4_days)][['datetime','PLa','PLb','PLc','PLtot']]
    pv_batch = pv[(today <= pv['datetime']) & (pv['datetime'] < next_4_days)][['datetime','PVa','PVb','PVc','PVtot']]
    
    # empty list for load_type and pv_type for each day
    load_type_list = []
    pv_type_list = []
    
    if load_batch.shape[0] == num_total_interval and pv_batch.shape[0] == num_total_interval:
        df_merge =  load_batch.merge(pv_batch,on = 'datetime',how = 'inner')    
        
        # create date_range for each batch    
        date_range = pd.date_range(start=today, periods=num_day, freq='D')

        # create load_type and pv_type for each day
        for day in date_range:
            type_load = df_load_type[df_load_type['date'] == day.date()]['load_type'].values[0]
            type_pv = df_pv_type[df_pv_type['date'] == day.date()]['solar_type'].values[0]
            
            load_type_list.append(type_load)
            pv_type_list.append(type_pv)
            
        data_small_batch = {'date': date_range,
                            'load_type': load_type_list,
                            'pv_type': pv_type_list}
        df_small_batch = pd.DataFrame(data_small_batch)
        
        load_count = df_small_batch['load_type'].value_counts()
        pv_count = df_small_batch['pv_type'].value_counts()
        
        load_type = load_count.idxmax()
        pv_type = pv_count.idxmax()
        
        # categorize batch dataset
        if load_type == 'high_load' and pv_type == 'high_solar':
            type1 += 1
            k = type1
            
        elif load_type == 'high_load' and pv_type == 'low_solar':
            type2 += 1
            k = type2
            
        elif load_type == 'medium_load' and pv_type == 'high_solar':
            type3 += 1
            k = type3
            
        elif load_type == 'medium_load' and pv_type == 'low_solar':
            type4 += 1
            k = type4
        
        elif load_type == 'low_load' and pv_type == 'high_solar':
            type5 += 1
            k = type5
        
        elif load_type == 'low_load' and pv_type == 'low_solar':
            type6 += 1
            k = type6  
        
        load_batch_mean = df_merge['PLtot'].mean()
        pv_batch_mean = df_merge['PVtot'].mean()
        
        df_merge.to_csv('data/batch_dataset_15min/' + load_type + '_' + pv_type + '_' + str(k) + '.csv',index=False)
        
        # prepare dataset_detail
        dataset_detail.append([load_type + '_' + pv_type + '_' + str(k), 
                               load_type, pv_type, 
                               load_batch_mean, pv_batch_mean, 
                               date_range[0], date_range[-1]])

# create dataframe of dataset_detail
dataset_detail = pd.DataFrame(dataset_detail, columns=['name','load_type','solar_type',
                                                      'load_batch_mean','pv_batch_mean',
                                                      'start_date', 'end_date'])
dataset_detail.to_csv('data/batch_dataset_15min/' + 'dataset_detail.csv',index=False)
dataset_detail
        

name    load_type  solar_type  load_batch_mean  \
0      high_load_high_solar_1    high_load  high_solar         5.429548   
1      high_load_high_solar_2    high_load  high_solar         5.838591   
2      high_load_high_solar_3    high_load  high_solar         9.243087   
3      high_load_high_solar_4    high_load  high_solar         8.144915   
4      high_load_high_solar_5    high_load  high_solar         8.136079   
..                        ...          ...         ...              ...   
254  medium_load_low_solar_45  medium_load   low_solar         4.657947   
255    high_load_low_solar_22    high_load   low_solar         4.939259   
256    high_load_low_solar_23    high_load   low_solar         4.490177   
257    high_load_low_solar_24    high_load   low_solar         7.168850   
258    high_load_low_solar_25    high_load   low_solar         6.333389   

     pv_batch_mean start_date   end_date  
0         1.032729 2023-03-16 2023-03-19  
1         1.061130 2023-03-17 2023-03-20  
2         1.147762 2023-03-22 2023-03-25  
3         1.152750 2023-03-23 2023-03-26  
4         1.155894 2023-03-24 2023-03-27  
..             ...        ...        ...  
254       0.451488 2024-02-07 2024-02-10  
255       0.441434 2024-02-15 2024-02-18  
256       0.476373 2024-02-16 2024-02-19  
257       0.513085 2024-02-21 2024-02-24  
258       0.526900 2024-02-22 2024-02-25  

[259 rows x 7 columns]

In [32]:
dataset_detail[['load_type', 'solar_type']].value_counts()

load_type    solar_type
low_load     low_solar     91
medium_load  low_solar     45
low_load     high_solar    39
high_load    high_solar    38
             low_solar     25
medium_load  high_solar    21
dtype: int64

# We have 259 batches
## ---------------------
## High Load & High Solar
- 38 batches

## High Load & Low Solar
- 12 batches
## ---------------------

## Low Load & High Solar
- 39 batches

## Low Load & Low Solar
- 58 batches

## ---------------------

## Medium Load & High Solar
- 22 batches

## Medium Load & Low Solar
- 36 batches


# 5 min resample

In [55]:
load = pd.read_csv('data/load_data_5minresample_concat.csv',parse_dates=[0]).rename(columns={'Pa (kW)':'PLa',
                                                                                         'Pb (kW)':'PLb',
                                                                                         'Pc (kW)':'PLc',
                                                                                         'Ptot (kW)':'PLtot'	
                                                                                         })
pv = pd.read_csv('data/pv_data_5minresample_concat.csv',parse_dates=[0]).rename(columns={'Pa (kW)':'PVa',	
                                                                                     'Pb (kW)':'PVb',
                                                                                     'Pc (kW)':'PVc',
                                                                                     'Ptot (kW)':'PVtot'	
                                                                                     })

In [56]:
start = max(load['datetime'].min(),pv['datetime'].min()) 
end = start + pd.Timedelta(9, unit='D')

In [57]:
#for 5 min sampling

num_day = 4 # unit : day
resolution = 5 # unit : min
num_interval_per_day = int(24*60/resolution)
num_total_interval = num_day * num_interval_per_day

# there are six types of batch dataset
type1 = 0
type2 = 0
type3 = 0
type4 = 0
type5 = 0
type6 = 0

start = max(load['datetime'].min(),pv['datetime'].min()) 
end = min(load['datetime'].max(),pv['datetime'].max())

dataset_detail = []
for i in range((end - start).days):
    today = start + pd.Timedelta(i,unit='D')
    next_4_days = today + pd.Timedelta(4,unit='D')
    load_batch = load[(today <=load['datetime']) & (load['datetime'] < next_4_days)][['datetime','PLa','PLb','PLc','PLtot']]
    pv_batch = pv[(today <= pv['datetime']) & (pv['datetime'] < next_4_days)][['datetime','PVa','PVb','PVc','PVtot']]
    
    # empty list for load_type and pv_type for each day
    load_type_list = []
    pv_type_list = []
    
    if load_batch.shape[0] == num_total_interval and pv_batch.shape[0] == num_total_interval:
        df_merge =  load_batch.merge(pv_batch,on = 'datetime',how = 'inner')    
        
        # create date_range for each batch    
        date_range = pd.date_range(start=today, periods=num_day, freq='D')

        # create load_type and pv_type for each day
        for day in date_range:
            type_load = df_load_type[df_load_type['date'] == day.date()]['load_type'].values[0]
            type_pv = df_pv_type[df_pv_type['date'] == day.date()]['solar_type'].values[0]
            
            load_type_list.append(type_load)
            pv_type_list.append(type_pv)
            
        data_small_batch = {'date': date_range,
                            'load_type': load_type_list,
                            'pv_type': pv_type_list}
        df_small_batch = pd.DataFrame(data_small_batch)
        
        load_count = df_small_batch['load_type'].value_counts()
        pv_count = df_small_batch['pv_type'].value_counts()
        
        load_type = load_count.idxmax()
        pv_type = pv_count.idxmax()
        
        # categorize batch dataset
        if load_type == 'high_load' and pv_type == 'high_solar':
            type1 += 1
            k = type1
            
        elif load_type == 'high_load' and pv_type == 'low_solar':
            type2 += 1
            k = type2
            
        elif load_type == 'medium_load' and pv_type == 'high_solar':
            type3 += 1
            k = type3
            
        elif load_type == 'medium_load' and pv_type == 'low_solar':
            type4 += 1
            k = type4
        
        elif load_type == 'low_load' and pv_type == 'high_solar':
            type5 += 1
            k = type5
        
        elif load_type == 'low_load' and pv_type == 'low_solar':
            type6 += 1
            k = type6  
        
        load_batch_mean = df_merge['PLtot'].mean()
        pv_batch_mean = df_merge['PVtot'].mean()
        
        df_merge.to_csv('data/batch_dataset_5min/' + load_type + '_' + pv_type + '_' + str(k) + '.csv',index=False)
        
        # prepare dataset_detail
        dataset_detail.append([load_type + '_' + pv_type + '_' + str(k), 
                               load_type, pv_type, 
                               load_batch_mean, pv_batch_mean, 
                               date_range[0], date_range[-1]])

# create dataframe of dataset_detail
dataset_detail = pd.DataFrame(dataset_detail, columns=['name','load_type','solar_type',
                                                      'load_batch_mean','pv_batch_mean',
                                                      'start_date', 'end_date'])
dataset_detail.to_csv('data/batch_dataset_5min/' + 'dataset_detail.csv',index=False)
dataset_detail
        

name  load_type  solar_type  load_batch_mean  \
0    high_load_high_solar_1  high_load  high_solar         5.431465   
1    high_load_high_solar_2  high_load  high_solar         5.837907   
2    high_load_high_solar_3  high_load  high_solar         9.241166   
3    high_load_high_solar_4  high_load  high_solar         8.144417   
4    high_load_high_solar_5  high_load  high_solar         8.136385   
..                      ...        ...         ...              ...   
221  low_load_high_solar_66   low_load  high_solar         4.252216   
222   low_load_low_solar_49   low_load   low_solar         1.939363   
223   low_load_low_solar_50   low_load   low_solar         1.809323   
224  low_load_high_solar_67   low_load  high_solar         1.395487   
225  low_load_high_solar_68   low_load  high_solar         1.026573   

     pv_batch_mean start_date   end_date  
0         1.032740 2023-03-16 2023-03-19  
1         1.061115 2023-03-17 2023-03-20  
2         1.147733 2023-03-22 2023-03-25  
3         1.152752 2023-03-23 2023-03-26  
4         1.155889 2023-03-24 2023-03-27  
..             ...        ...        ...  
221       0.538947 2023-12-18 2023-12-21  
222       0.496592 2023-12-25 2023-12-28  
223       0.520878 2023-12-26 2023-12-29  
224       0.530873 2023-12-27 2023-12-30  
225       0.520201 2023-12-28 2023-12-31  

[226 rows x 7 columns]

In [58]:
dataset_detail[['load_type', 'solar_type']].value_counts()

load_type    solar_type
low_load     high_solar    68
             low_solar     50
high_load    high_solar    38
medium_load  low_solar     30
             high_solar    28
high_load    low_solar     12
dtype: int64